Issue 1268: Fix lazy slicing stopping at the first deleted

jstuder-gh opened the pull request:
This commit fixes the issue in which using lazy Iterables as Positional
indices will evaluate until the first deleted (undefined) value but not
continue further.

Within the default eagerize routine, it checks the number of reified
elements upon encountering an undefined value to determine whether to
continue.

Addresses [Issue #1268](https://github.com/rakudo/rakudo/issues/1268)

jstuder-gh commented:
Will need to write tests if accepted.

zoffixznet commented:
Is there a way we can we do this without adding a public method?

I also don't understand how this fixes anything for Seqs, since reified part will often just contain whatever was already reified:

```
<Zoffix__> m: with [<a b c d e>] { .[0]:delete; with .Seq { use nqp; nqp::elems( nqp::getattr(.cache, List, '$!reified') ).say } }
<camelia> rakudo-moar 1dbf5f589: OUTPUT: «0␤»
```

And so, IMO this type of fix just muddles up the conditions where a hole would be stopped at. "Stops at first gap or end of Iterable" is easy to understand. But if you also have to care about the kind of iterable you, as a user, have you now have 4 different cases to consider instead of just 2.

IMO we should either fix this entirely or not fix it at all (and document behaviour as described in the [Issue](https://github.com/rakudo/rakudo/issues/1268#issuecomment-347023393)). Doing half-way measures doesn't improve usability or usefulness of the feature.

jstuder-gh commented:
I added a commit that addresses the issue of pending, unreified elements on the List or Seq as showcased in your example by explicitly reifying until lazy if the Reifier indicates that there is more to do.

As for the public methods, I suppose we could encapsulate all of this logic within the eagerize block. I had just figured this was a bit cleaner. But if the number of reified elements is not something we want to be exposing to the user, I can get rid of them.

zoffixznet commented:
> But if the number of reified elements is not something we want to be exposing to the user

No we shouldn't, especially not if calling the method *causes*, often complete, reification.

------

BTW is the `Seq` handling needed at all? Looking more at it now and it looks like the hole is filled when an Array is converted to Seq/List and this issue only exists with Arrays, since you can't `:delete` from Seqs/Lists directly:

```
<Zoffix__> m: my @a2 = 4, 5, 6; @a2[1]:delete; say @a2.Seq[lazy 0, 1, 2, 3, 4, 5, 6, 7]
<camelia> rakudo-moar 1dbf5f589: OUTPUT: «(4 (Any) 6)␤»

<Zoffix__> m: my @a2 = 4, 5, 6; @a2[1]:delete; say @a2.List[lazy 0, 1, 2, 3, 4, 5, 6, 7]
<camelia> rakudo-moar 1dbf5f589: OUTPUT: «(4 Nil 6)␤»
```

jstuder-gh commented:
> No we shouldn't, especially not if calling the method causes, often complete, reification.

Good point. I'll remove the methods and incorporate the logic into the eagerizer in the next iteration. I'll also try removing the Seq handling as well.

jstuder-gh commented:
As discussed, I've removed the public methods and incorporated the logic into the default "eagerize" block for POSITIONS. It also passes the spectest.

zoffixznet commented:
I don't see anyone else commenting, but I'm still against this sort of piece-meal fixes. All it's doing is making the behaviour hard to predict.

(the latest version of this PR assumes anything that isn't a Seq is a List, which will make this codepath explode for userland Positionals)

jstuder-gh commented:
I'm closing this PR as it's clear now that this was not the right path to take in order to fix this problem. There is still discussion at [Issue 1268](https://github.com/rakudo/rakudo/issues/1268) about another approach. If that proves to be the right way forward, I will open another PR at a later date.
